### ISSUE TITLE
Add ffmpeg bundle

### ISSUE THREAD

**Issue by fabiohl** [clearlinux/distribution#429]
Clear is great choice nor only for server or desktops, but also for automated build machines. For example, a spool of encoding videos, compiling executables, and others. Its focus on performance would a natural choice for encoding multimedia.
So I strongly suggest a bundle for ffmpeg and all codecs.

**Comment by nesiusra** [comment]
@chivakker - related to some of your long term initiatives?

**Comment by chivakker** [comment]
Yes, this is work in progress.  But nothing should prevent a user to compile properly ffmpeg by following their own guide to include whatever library is appropriate.  I can certainly help to figure that out if the user shares their configure line and errors.

**Comment by danboid** [comment]
@ahkok I'm very concerned you blocked this request because ffmpeg is the number one missing package in Clear Linux for me at present and is the root cause of why I won't be running Clear as my desktop OS until ffmpeg gets bundled.

GnomeMPV is not a substitute for ffmpeg. Whilst mpv is my video player of choice, ffmpeg is mainly used for encoding and transcoding which (Gnome)MPV doesn't handle. The main reason I require ffmpeg support as part of my desktop Linux OS is for ffmpegthumbnailer which is used by tumbler / Thunar to generate thumbnail previews of video files. I am going to open a separate ticket for that because it was the killer blow that has sent me packing until Clear has better media playback support, to which ffmpeg is absolutely key.

**Comment by ahkok** [comment]
I'm not blocking this request. Legal issues are blocking this request.

**Comment by nesiusra** [comment]
> /usr/bin/ld: libavcodec/mqc.o: relocation R_X86_64_32 against `.rodata' can not be used when making a shared object; recompile with -fPIC
> /usr/bin/ld: libavcodec/mqcdec.o: relocation R_X86_64_32S against undefined symbol `ff_mqc_qe' can not be used when making a shared object; recompile with -fPIC
> /usr/bin/ld: libavcodec/svq13.o:

@gabrielgiordano - the problem is that those three object files were not compiled with -fPIC.  Make sure -fPIC is in your CFLAGS/CXXFLAGS.

**Comment by ahkok** [comment]
The `-fPIC` breakage is new, I didn't have this error before, and it does break new compiles.

**Comment by nesiusra** [comment]
Then it's more likely CFLAGS or something like it isn't being given to the $CC lines for those files.

**Comment by pceiley** [comment]
Does compiling ffmpeg still fail with the issue above?

**Comment by pceiley** [comment]
I can make ffmpeg without errors, but I get the following when attempting to execute it:
```
# ffmpeg 
ffmpeg: error while loading shared libraries: libavdevice.so.58: cannot open shared object file: No such file or directory
```
Any suggestions?

**Comment by gvancuts** [comment]
I guess I'd start by checking whether the `libavdevice.so.58` file is in your rootfs.

Did you build `ffmpeg` on the very same machine you're trying to run it on?

**Comment by ahkok** [comment]
After compiling `ffmpeg`, you need to run `ldconfig` as root once.

**Comment by miguelinux** [comment]
> After compiling `ffmpeg`, you need to run `ldconfig` as root once.
> 

And update /etc/ld.so.conf with the new paths of your libraries

**Comment by pceiley** [comment]
Thanks for the help all - overlooked this in the ffmpeg doc! Sorry to be a bother.

**Comment by ahkok** [comment]
We've made significant progress with the various aspects of this package - this is progressing well and above my expectations.

**Comment by pceiley** [comment]
Any hint on what the solution might look like?

**Comment by ahkok** [comment]
Tentatively: We're building a version of `ffmpeg` that fully relies on video HW decode/encode capabilities in modern Intel Hardware (APL NUC, SKL NUC and newer) and doesn't have any software codecs that we can not ship. `vlc` will be the first package to use it. `ffplay` should also be able to use it.

Some older Intel HW may not be able to display all (video/audio) content, and there will still be content that may give issues, but it should be a significant step forward.

**Comment by pceiley** [comment]
Thanks. Appreciate the insight.

**Comment by eero-t** [comment]
> Some older Intel HW may not be able to display all (video/audio) content, and there will still be content that may give issues, but it should be a significant step forward.

You could provide both "i965_drv_video.so" & "iHD_drv_video.so" libva backends.  Former supports HW decode/encode at least for GEN6/7/8 (and limited set of GEN9 capabilities), latter supports BDW and GEN9+ HW.

Then you need something that selects suitable backend (e.g. by setting LIBVA_DRIVER_NAME either to "iHD" or "i965") based on HW generation ("sudo awk '/^gen:/ {print $2}' /sys/kernel/debug/dri/0/i915_capabilities").

**Comment by danboid** [comment]
What @eero-t said. Please don't exclude older but still capable hardware.

I had no luck getting quicksync encoding working under FFMPEG last time I tried so hopefully this work will make using QSV easier/ more reliable, at least under Clear.

**Comment by eero-t** [comment]
FFmpeg VAAPI backed is a bit easier to use as you don't need to specify codec when decoding, only when encoding.  QSV (i.e. MediaSDK) is mostly just a wrapper over libva, offering Windows like API for the same functionality (there's some extra functionality compared to VAAPI, but not much).

Here are FFmpeg HEVC -> 6Mbits AVC transcoding examples for both backends:
* QSV: `ffmpeg  -y -hwaccel qsv -qsv_device /dev/dri/renderD128 -c:v hevc_qsv -i input.h265 -c:v h264_qsv -b:v 6M  output.h264`
* VAAPI: `ffmpeg  -y -hwaccel vaapi -vaapi_device /dev/dri/renderD128 -hwaccel_output_format vaapi -i input.h265 -c:v h264_vaapi -b:v 6M  output.h264`

(In case you have multiple GPU device, e.g. Intel HadesCanyon NUC, you need to specify correct device to use also for QSV, for some reason underlying VAAPI stuff defaults to same device as X server, although that doesn't necessary have anything to do with video acceleration.)

**Comment by pceiley** [comment]
I've built ffmpeg with libx264, but I'm having difficulties getting Firefox to load the libraries. They are located in /usr/loca/lib and I've added to /etc/ld.so.conf, run ldconfig and confirmed it's been added using ldconfig -p.

If I execute `LD_LIBRARY_PATH=/usr/local/lib firefox` then it works correctly, bur otherwise Firefox doesn't seem to honour the ldconfig cache. Has anyone else had this issue?

**Comment by thiagomacieira** [comment]
Can you confirm with `ldd` on the actual library/plugin in question and see if /etc/ld.so.conf applied? If that failed, can you try a file in /etc/ld.so.conf.d/ ?

**Comment by ahkok** [comment]
/usr/local/lib is not by default added to /etc/ld.so.conf*

**Comment by thiagomacieira** [comment]
No, it isn't, but the point is that @pceiley added it to /etc/ld.so.conf and that didn't seem to take effect.

**Comment by OpenSourceAnarchist** [comment]
Just out of curiosity, could Red Hat, Cisco, Endless, and Centricular's efforts be put to use in Clear with their improvements to OpenH264 (as per [this blog post](https://blogs.gnome.org/uraeus/2019/04/03/preparing-for-fedora-workstation-30/)?

Something to look into for Firefox, perhaps. Is the new ffmpeg solution in Clear compatible with Haswell processors or does it have to be Broadwell and newer?

**Comment by pceiley** [comment]
TL;DR
- adding /usr/local/lib to /etc/ld.so.conf seems to work as expected (but not with Firefox!).
- creating a .conf file under /etc/ld.so.conf.d has no effect. Do I need to create a /etc/ld.so.conf file that sources files in /etc/ld.so.conf.d/*?

With nothing in /etc/ld.so.conf:
```
$ ldconfig -p |grep /usr/local
$
$ ldd /usr/local/lib/libavcodec.so
	linux-vdso.so.1 (0x00007ffcaf0f9000)
	libswresample.so.3 => not found
	libavutil.so.56 => not found
	libm.so.6 => /usr/lib64/haswell/libm.so.6 (0x00007f71f1c80000)
	libz.so.1 => /usr/lib64/haswell/libz.so.1 (0x00007f71f1c57000)
	libx264.so.155 => not found
	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007f71f1c35000)
	libc.so.6 => /usr/lib64/haswell/libc.so.6 (0x00007f71f1a43000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007f71f3261000)
```
With /usr/local/lib added to /etc/ld.so.conf:
```
$ ldconfig -p |grep /usr/local
	libx264.so.155 (libc6,x86-64) => /usr/local/lib/libx264.so.155
	libx264.so (libc6,x86-64) => /usr/local/lib/libx264.so
	libswscale.so.5 (libc6,x86-64) => /usr/local/lib/libswscale.so.5
	libswscale.so (libc6,x86-64) => /usr/local/lib/libswscale.so
	libswresample.so.3 (libc6,x86-64) => /usr/local/lib/libswresample.so.3
	libswresample.so (libc6,x86-64) => /usr/local/lib/libswresample.so
	libpostproc.so.55 (libc6,x86-64) => /usr/local/lib/libpostproc.so.55
	libpostproc.so (libc6,x86-64) => /usr/local/lib/libpostproc.so
	libavutil.so.56 (libc6,x86-64) => /usr/local/lib/libavutil.so.56
	libavutil.so (libc6,x86-64) => /usr/local/lib/libavutil.so
	libavformat.so.58 (libc6,x86-64) => /usr/local/lib/libavformat.so.58
	libavformat.so (libc6,x86-64) => /usr/local/lib/libavformat.so
	libavfilter.so.7 (libc6,x86-64) => /usr/local/lib/libavfilter.so.7
	libavfilter.so (libc6,x86-64) => /usr/local/lib/libavfilter.so
	libavdevice.so.58 (libc6,x86-64) => /usr/local/lib/libavdevice.so.58
	libavdevice.so (libc6,x86-64) => /usr/local/lib/libavdevice.so
	libavcodec.so.58 (libc6,x86-64) => /usr/local/lib/libavcodec.so.58
	libavcodec.so (libc6,x86-64) => /usr/local/lib/libavcodec.so
$ ldd /usr/local/lib/libavcodec.so
	linux-vdso.so.1 (0x00007ffed73dd000)
	libswresample.so.3 => /usr/local/lib/libswresample.so.3 (0x00007fb409bb5000)
	libavutil.so.56 => /usr/local/lib/libavutil.so.56 (0x00007fb409b31000)
	libm.so.6 => /usr/lib64/haswell/libm.so.6 (0x00007fb4099ee000)
	libz.so.1 => /usr/lib64/haswell/libz.so.1 (0x00007fb4099c5000)
	libx264.so.155 => /usr/local/lib/libx264.so.155 (0x00007fb4096b0000)
	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007fb40968e000)
	libc.so.6 => /usr/lib64/haswell/libc.so.6 (0x00007fb40949c000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007fb40b072000)
	libdl.so.2 => /usr/lib64/libdl.so.2 (0x00007fb409496000)
```
Now, if I mv /etc/ld.so.conf to /etc/ld.so.conf.d/usrlocal.conf:
```
$ ldconfig -p |grep /usr/local
$
$ ldd /usr/local/lib/libavcodec.so
	linux-vdso.so.1 (0x00007ffdcfb87000)
	libswresample.so.3 => not found
	libavutil.so.56 => not found
	libm.so.6 => /usr/lib64/haswell/libm.so.6 (0x00007fe866475000)
	libz.so.1 => /usr/lib64/haswell/libz.so.1 (0x00007fe86644c000)
	libx264.so.155 => not found
	libpthread.so.0 => /usr/lib64/libpthread.so.0 (0x00007fe86642a000)
	libc.so.6 => /usr/lib64/haswell/libc.so.6 (0x00007fe866238000)
	/usr/lib64/ld-linux-x86-64.so.2 (0x00007fe867a56000)
```

**Comment by lebensterben** [comment]
I've built ffmpeg a while ago and also add it to ld.so.conf, but it doesn't work for Firefox.

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by eero-t** [comment]
> creating a .conf file under /etc/ld.so.conf.d has no effect. Do I need to create a /etc/ld.so.conf file that sources files in /etc/ld.so.conf.d/*?

Naturally.  Most other distros have this, but ClearLinux doesn't:
`include /etc/ld.so.conf.d/*.conf`

**Comment by eero-t** [comment]
> Just out of curiosity, could Red Hat, Cisco, Endless, and Centricular's efforts be put to use in Clear with their improvements to OpenH264 (as per this blog post)?

Why?  MediaSDK / iHD media-driver already support H264 main & high profiles.

> Something to look into for Firefox, perhaps. Is the new ffmpeg solution in Clear compatible with Haswell processors or does it have to be Skylake and newer?

MediaSDK / iHD media-driver support Broadwell and newer.  The old i965 VAAPI driver supports Skylake.  But I don't know whether latter was also added to ClearLinux.

**Comment by mariojuniorjp** [comment]
> There is nothing special about compiling ffmpeg on clearlinux. The last time I did it I just downloaded it and did `configure` and `make`.

Now try to compile with libx264 and libx265. Good luck!

**Comment by pceiley** [comment]
For those having trouble, here's how I compiled ffmpeg for use with Firefox.

You'll need the C compiler installed

```
sudo swupd bundle-add c-basic
```
Now create a build directory somewhere and cd into it.
```
mkdir ~/build && cd ~/build
```
Install ffmpeg
```
FFMPEG_PKGNAME=ffmpeg-4.1.3.tar.bz2
wget https://ffmpeg.org/releases/$FFMPEG_PKGNAME
tar jxf $FFMPEG_PKGNAME
cd $FFMPEG_PKGNAME
./configure --prefix=/usr/local --enable-shared && make && sudo make install
```
If you haven't already added /usr/local/lib to ld.so.conf, this is required before ffmpeg will work.
```
sudo sh -c 'echo /usr/local/lib >>/etc/ld.so.conf'
sudo ldconfig
```
Now this **should** be enough to make Firefox reference the new libraries, but for some reason it doesn't. I suspect it might be an issue with Firefox upstream.

The below is an awful, awful hack and should be avoided, but until someone has a better solution...
```
sudo sh -c 'echo LD_LIBRARY_PATH=/usr/local/lib >>/etc/environment'
```
Now reboot your system and verify Firefox by visiting [https://www.youtube.com/html5](https://www.youtube.com/html5)

**Comment by shelterx** [comment]
> The below is an awful, awful hack and should be avoided, but until someone has a better solution...
> 
> ```
> sudo sh -c 'echo LD_LIBRARY_PATH=/usr/local/lib >>/etc/environment'
> ```

Tricky one, works for me also, thank you.

**Comment by shelterx** [comment]
It worked until I rebooted, not sure what broke it.
If I start firefox from bash with "LD_LIBRARY_PATH=/usr/local/lib ~/firefox/firefox-bin" it works.

**Comment by miguelinux** [comment]
> ```
> sudo sh -c 'echo LD_LIBRARY_PATH=/usr/local/lib >>/etc/environment'
> ```

On CL, firefox is loaded by the script /usr/bin/firefox, which said that
it cat take user config from 

USER_CONFIG="${HOME}/.config/firefox.conf" 

So you can do:

echo "export LD_LIBRARY_PATH=/usr/local/lib >> ${HOME}/.config/firefox.conf

**Comment by pceiley** [comment]
Works a treat - thanks so much @miguelinux. I'm going to update my post above accordingly as messing around with LD_LIBRARY_PATH system wide is considered bad practice and I don't want people breaking their systems.

Do you think the issue is upstream with Firefox? If so, I'll log a bug - but won't hold my breath :)

PS. missing closing " above.

**Comment by shelterx** [comment]
Even you edit the ${HOME}/.config/firefox.conf file. you still need the /usr/local/lib entry in /etc/ld.so.conf otherwise it won't work for some reason.
Noticed that after a reinstall.

**Comment by mariojuniorjp** [comment]
OMG, I love you guys! Now MP4 and GIF files are working on Firefox. :+1:

**Comment by mariojuniorjp** [comment]
Speaking of which: Has anyone managed to generate thumbnails in the gnome?

**Comment by ahkok** [comment]
#885 for tracking video thumbnails in GNOME.

**Comment by ahkok** [comment]
I'm closing this issue. Note the feature is implemented in the `vlc` and `mpv` bundles.

For future *increases* in functionality, please file *new* requests (e.g. for missing codecs specifically, muxers/demuxers) so that we can solve them on a per-case basis.

**Comment by shelterx** [comment]
ffmpeg is still needed for video playback in firefox tho'

**Comment by ahkok** [comment]
We can't ever offer you an ffmpeg that includes `libx264` and other software codecs that are patented. For that reason, it makes no sense to keep this issue open forever. Sorry.

**Comment by OpenSourceAnarchist** [comment]
@ahkok What about native deciding through Intel's "media-driver" VAAPI software? Could that somehow be used to allow hardware decoding/encoding for general applications, including Firefox? Don't really need software codecs when hardware can be directly used.

**Comment by ahkok** [comment]
We can HW decode the video stream, but not `AAC` - so no matter what we can't make many web videos work out of the box right now.

**Comment by OpenSourceAnarchist** [comment]
Too bad, but thanks for explaining. Seems like no real good solution. I'd really like to switch to Clear for desktop use but unfortunately just Arch and Gentoo provide the functionality that I need for video encoding/decoding. I understand the legal issues but it's amazing that other distros just include the software anyway. I suppose Intel must be more careful with licensing.

**Comment by ahkok** [comment]
I'm not stating there is zero solutions.

This issue thread is overly long and no longer *specific*. We encourage everyone to open *specific* new issues for new additions so we can discuss and come up with working solutions for them as much as possible (even for `AAC` we already have some ideas that may make it easier in the future ...).

So, let's keep this thread closed and move discussion to new threads on specific subtopics.
